### DISCUSSION TITLE
Missing `reportUnnecessaryComparison` when comparing objects of different types

### DISCUSSION THREAD

**Discussion by tkukushkin** [microsoft/pyright#9349] [Q&A] [answered]
Hello! I've noticed, that pyright doesn't report `reportUnnecessaryComparison` when I compare two objects of different types with `is` operator:

```python
class Foo:
    pass

def test(x: list[Foo], y: Foo) -> None:
    if x is y:
        pass
```

I understand why this might not work with `==`, but I think this should work with `is`. It works with `int` for example.

**Comment by erictraut** [comment]
What you're suggesting isn't safe. It's possible to create a class that subclasses from both `Foo` and `list`.

```python
class Bar(Foo, list):
    pass

b = Bar()
test(b, b)
```

The `reportUnnecessaryComparison` check must be rather conservative to avoid false positive errors. This is true even for `is` and `is not` narrowing.

**Reply by tkukushkin** [reply]
But is pyright able to understand that Foo is not list as it seems it does with `int` and `list[int]`?

**Reply by erictraut** [reply]
I'm not sure what you mean. Could you provide a code sample?

**Reply by tkukushkin** [reply]
```python
def test(x: list[int], y: int) -> None:
    if x is y:
        pass
```
Gives `Condition will always evaluate to False since the types "list[int]" and "int" have no overlap  (reportUnnecessaryComparison)`

Is it so difficult to derive that `Foo` doesn't inherit from `list`? Or am I looking at the problem too narrowly and realising that types don't overlap in the general case is too difficult?

**Reply by erictraut** [reply, marked as answer]
There are some heuristics involved in pyright's check here. It assumes that if the types are both built-in types like `int` and `list`, then it's not possible to create a subclass of the two types. Indeed, if you try to create a class that subclasses from both `int` and `list`, you will see that the runtime doesn't allow this. Compare this to my example above where it is possible to create a subclass of both `Foo` and `list`.

All else equal, when we need to choose between a potential false positive and a potential false negative, we tend to opt for the false negative. That informs the heuristics in this check.

In any case, pyright is working as intended here.

**Reply by tkukushkin** [reply]
Got it, thanks!
